phpBB/phpbb/template/twig/node/event.php

```diff
@@ -46,80 +46,49 @@ public function compile(\Twig\Compiler $compiler)
 
 		$location = $this->listener_directory . $this->getNode('expr')->getAttribute('name');
 
-		$compiler_steps = [];
+		$template_events = [];
 
+		// Group and sort extension template events in according to their priority (0 by default if not set)
 		foreach ($this->environment->get_phpbb_extensions() as $ext_namespace => $ext_path)
 		{
 			$ext_namespace = str_replace('/', '_', $ext_namespace);
+			$priority_key = $this->template_event_priority_array[$ext_namespace][$location] ?? 0;
+			$template_events[$priority_key][] = $ext_namespace;
+		}
+		krsort($template_events);
```

Comment:
As discussed in discord I will move the `if ($this->environment->isDebug() || $this->environment->getLoader()->exists('@' . $ext_namespace . '/' . $location . '.html'))` here myself after I approve this PR.

Without that if statement here, $template_events gets filled up with too many unnecessary calls for events that don't exist. With it here, it executes as it already does in phpBB, so it should be here.

Just making this comment for archive's sake and to have a note of it.

Reply:
Unresolving this temporarily just to make a comment to flesh out the logic of how this stuff works, as it sometimes confuses me when thinking about it.

If `$priority_key` is already set, then it will append the `$ext_namespace` to the end of that `$priority_key` array.

As written, technically, the `$events` [on line 62](https://github.com/toxyy/phpbb/blob/0737f16c068494dd27eea47fa18cc4a3402a670b/phpBB/phpbb/template/twig/node/event.php#L60) in `foreach ($template_events as $events)` should only be of size 1. It will only ever be of a bigger size if something else has the same `$priority_key`, which means it was manually set to an existing one. As I initially intended in my PR on phpBB, if setting to the same `$priority_key`, everything should be shifted up by 1 accordingly, as it would take that key. Since the inner arrays are not being sorted, it will work as intended. `$events` will be sorted in order of being added, so something with the same key will be appended to the end, meaning it will execute last. So this is intended behavior and is fine.

Just needed to write this out since I got confused, in case I need to reference it in the future.